### PR TITLE
Improve the Spark blueprint

### PR DESCRIPTION
**Pull request by kayousterhout**
This PR includes a bunch of fixes to cleanup the Spark blueprint

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -12,6 +12,14 @@ RUN apt-get update && apt-get install -y \
 && mv /spark* /spark \
 && rm -rf /var/lib/lists/* /tmp/* /var/tmp/*
 
+# Create the directory where the Spark event log, which stores information about
```
Comment (ejj) [inline Dockerfile]: This first sentence is a bit hard to parse.  Perhaps just "Create a directory for the Spark event log, which stores information about past jobs that have completed."

File: spark.js [hunk 2]
```diff
@@ -27,10 +28,25 @@ function Spark(nWorker) {
...
     '/spark/conf/spark-defaults.conf': sparkConfContent,
+    '/spark/conf/log4j.properties': log4jConfigContent,
   };
+  return sparkConfigFiles;
+}
+
+/**
+ * Spark creates a Spark cluster (a master and a collection of workers).
+ *
+ * @param {number} nWorker The number of workers to boot.
```
Comment (ejj) [inline spark.js]: Maybe just "the number of workers" in the comment.  Booting sounds not precisely accurate.

File: README.md [hunk 3]
```diff
@@ -42,51 +42,93 @@ Next, run the Spark blueprint:
 $ kelda run ./sparkRun.js
 ```
 
-This will use the blueprint `sparkRun.js` to launch a cluster of Spark workers
-and a Spark master. It will take a bit for the VMs to boot up, for Kelda to
+This will use the blueprint `sparkRun.js` to launch a Spark cluster. It will take a
```
Comment (ejj) [inline README.md]: Did you intend to make all of these readme updates in this patch about the driver?  They seem unrelated.  Perhaps a rebasing error?

Comment (ejj) [reply README.md]: Ah nvm I think it's fine.

File: spark.js [hunk 4]
```diff
@@ -42,72 +42,113 @@ function getConfigFiles() {
...
-    // XXX: This is only necessary so that spark nodes can ask an external
-    // service what their public IP is.  Once this information can be passed in
-    // through an environment variable, these ACLs should be removed.
-    publicInternet.allowFrom(this.workers, 80);
-    publicInternet.allowFrom(this.master, 80);
+    // XXX: This is only necessary so that the driver can learn its public IP
+    // (see note above).
+    publicInternet.allowFrom(this.driver, 80);
+
+    console.log('Spark driver started with hostname "spark-driver". ' +
```
Comment (ejj) [inline spark.js]: I think in the log message it might be better to query the container for what it's hostname is.   If it's the second driver created, it would be called `spark-driver1` for example ...

Also that's not a "hostname"  so much as it's a "name".  The hostname will actually be something like spark-driver.q

File: spark.js [hunk 5]
```diff
@@ -47,8 +47,12 @@ class Spark {
    * run a Spark application.
    *
    * @param {number} nWorker The number of workers to boot.
+   * @param {string} [appImage=image] Image that should be used for the Spark
+   *   Driver container, which manages running a particular application.
+   *   Specify a custom image here if you'd like to run a custom Spark
+   *   application.
    */
-  constructor(nWorker) {
+  constructor(nWorker, appImage = image) {
```
Comment (ejj) [inline spark.js]: I like this is a lot better.

Do you still think we need the setImage() function and the global image variable?  We could just say `appImage = keldaio/spark` here for example.

File: spark.js [hunk 6]
```diff
@@ -83,17 +87,19 @@ class Spark {
...
    * Adds a container that users can use to run Spark jobs (e.g., by launching
    * spark-shell).
    *
+   * @param {string} imageUri - The URI of a Docker image that should be used
+   *   as the Driver's container.
    * @returns {void}
    */
-  addDriver() {
-    this.driver = new Container('spark-driver', image, {
+  addDriver(imageUri) {
```
Comment (ejj) [inline spark.js]: Did you consider in the constructor doing `this.image = appImage` so that here you could just query it without passing it in as an argument again?

File: README.md (hunk 3 above)
Comment (kayousterhout) [reply README.md]: Yeah as it sounds like you figured out, I updated a bunch of text about the Driver and then did a few other minor changes that I ran into while doing that.

File: spark.js (hunk 4 above)
Comment (kayousterhout) [reply spark.js]: Right now nothing will work if a second driver is created (e.g., none of the ports are setup correctly).  Do you still think we should query for the name?

Comment (kayousterhout) [reply spark.js]: Also, in the `kelda show` output, this is in a column labeled `HOSTNAME`, so we use hostname in a bunch of places to refer to this name.  Even though it's not totally accurate I think it's helpful that it's consistent with what we use elsewhere.  Do you think we should change this everywhere in the Kelda codebase to be something like "name" instead of "hostname"?

Comment (ejj) [reply spark.js]: If it's easy I think doing so is future proof.  It's just like `this.driver.hostname` or something right?  Anything more complicated than that and I agree it's not worth it.

File: spark.js (hunk 5 above)
Comment (kayousterhout) [reply spark.js]: Right now there are two images: `setImage` sets the image for all of the Spark containers _except_ the application (i.e., the driver container), and this is the one just for the container.  So, I still used `setImage` for testing against my own compiled image.

Also, gnocchi can't yet use this unfortunately, because it relies on local files that need to be copied to all of the workers (and those files are too big to include using `filepathToContent` in the blueprint).  The "right" way of doing things would be for that application to read the test data from HDFS rather than from the local filesystem, but we can't do that yet since we don't yet have HDFS.

This is all to say that this change is anticipating future simplicity, but isn't actually useful yet.

File: spark.js (hunk 6 above)
Comment (kayousterhout) [reply spark.js]: That seems a little heavy weight to me considering that it's just used once -- to make that change, the code in the main constructor would change to:

```
this.appImage = appImage;
this.addDriver();
```

That seems like it makes the code a little harder to read because it's not obvious why we're setting appImage / how it's used.

File: spark.js (hunk 5 above)
Comment (ejj) [reply spark.js]: Ok I misunderstood this then.

I wonder if this would be even simpler if we just required everyone to use the same image, and set that image through the constructor.  I can see perhaps some small container size advantages to allowing the application containers to have a different image than the driver or the master, but is it worth the additional complexity?  Having to ways to set the image may be confusing.

What do you think?

Comment (kayousterhout) [reply spark.js]: I removed this commit from the PR for now; let's discuss in person tomorrow b/c there are some subtleties to this

**Comment by kayousterhout** [comment]
Updated this and removed the image-setting commit
